File: src/main/java/co/gatedaccess/web/model/SecurityGuardOtp.java
package co.gatedaccess.web.model;

import com.mongodb.lang.NonNull;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.Field;
import org.springframework.data.mongodb.core.mapping.MongoId;

import java.util.Date;

@Document("security_guard_otp")
public class SecurityGuardOtp {
    @MongoId
    String id;

    @NonNull
    @Indexed(unique = true)
    String code;

    @NonNull
    @Indexed(unique = true)
    String communityId;

    @NonNull
    @Field("expire_at")
    Date expireAt;

    public SecurityGuardOtp() {
    }

    @CreatedDate
    @Field("created_at")
    Date createdAt;

    public SecurityGuardOtp(@NonNull String communityId, @NonNull Date expireAt, @NonNull String code) {
        this.communityId = communityId;
        this.expireAt = expireAt;
        this.code = code;
    }

    public String getId() {
        return id;
    }

    public String getCode() {
        return code;
    }

    public String getCommunityId() {
        return communityId;
    }

    public Date getCreatedAt() {
        return createdAt;
    }

    @NonNull
    public Date getExpireAt() {
        return expireAt;
    }

    public void setCode(@NonNull String code) {
        this.code = code;
    }

    public void setCommunityId(@NonNull String communityId) {
        this.communityId = communityId;
    }

    public void setExpireAt(@NonNull Date expireAt) {
        this.expireAt = expireAt;
    }

    public static final class Builder {
        private Date expireAt;
        private String communityId;
        private String code;

        public Builder() {
        }

        public static Builder aSecurityGuardOtp() {
            return new Builder();
        }

        public Builder withExpireAt(Date expireAt) {
            this.expireAt = expireAt;
            return this;
        }

        public Builder withCommunityId(String communityId) {
            this.communityId = communityId;
            return this;
        }

        public Builder withCode(String code) {
            this.code = code;
            return this;
        }

        public SecurityGuardOtp build() {
            SecurityGuardOtp securityGuardOtp = new SecurityGuardOtp();
            securityGuardOtp.communityId = this.communityId;
            securityGuardOtp.expireAt = this.expireAt;
            securityGuardOtp.code = this.code;
            return securityGuardOtp;
        }
    }
}

File: src/main/kotlin/co/gatedaccess/web/data/model/SecurityGuardOtp.kt
package co.gatedaccess.web.data.model

import com.mongodb.lang.NonNull
import jakarta.validation.constraints.Future
import org.springframework.data.annotation.CreatedDate
import org.springframework.data.mongodb.core.index.Indexed
import org.springframework.data.mongodb.core.mapping.Document
import org.springframework.data.mongodb.core.mapping.Field
import org.springframework.data.mongodb.core.mapping.MongoId
import java.util.*

@Document("security_guard_otp")
class SecurityGuardOtp {
    @MongoId
    var id: String? = null

    @JvmField
    @NonNull
    @Indexed(unique = true)
    var code: String? = null

    @JvmField
    @NonNull
    @Indexed(unique = true)
    var communityId: String? = null

    @JvmField
    @NonNull
    @Field("expire_at")
    var expireAt: @Future Date? = null

    constructor()

    @CreatedDate
    @Field("created_at")
    var createdAt: Date? = null

    constructor(
        id: String?,
        @NonNull code: String?,
        @NonNull communityId: String?,
        @NonNull expireAt: Date?,
        createdAt: Date?
    ) {
        this.id = id
        this.code = code
        this.communityId = communityId
        this.expireAt = expireAt
        this.createdAt = createdAt
    }

    fun getCode(): String? {
        return code
    }

    fun getCommunityId(): String? {
        return communityId
    }

    @NonNull
    fun getExpireAt(): Date? {
        return expireAt
    }

    fun setCode(@NonNull code: String?) {
        this.code = code
    }

    fun setCommunityId(@NonNull communityId: String?) {
        this.communityId = communityId
    }

    fun setExpireAt(@NonNull expireAt: Date?) {
        this.expireAt = expireAt
    }

    class Builder {
        private var expireAt: Date? = null
        private var communityId: String? = null
        private var code: String? = null

        fun withExpireAt(expireAt: Date?): Builder {
            this.expireAt = expireAt
            return this
        }

        fun withCommunityId(communityId: String?): Builder {
            this.communityId = communityId
            return this
        }

        fun withCode(code: String?): Builder {
            this.code = code
            return this
        }

        fun build(): SecurityGuardOtp {
            val securityGuardOtp = SecurityGuardOtp()
            securityGuardOtp.communityId = this.communityId
            securityGuardOtp.expireAt = this.expireAt
            securityGuardOtp.code = this.code
            return securityGuardOtp
        }

        companion object {
            fun aSecurityGuardOtp(): Builder {
                return Builder()
            }
        }
    }
}
